### PR TITLE
refactor(ApplicationAutoscaling) - move logic to static functions

### PR DESCRIPTION
**Pull request by jpetto**
# Goal

move logic to static functions.

- move code out of constructor into static functions
- make code a bit more DRY
- add tests for all static functions
- add a test helper for an empty/blank construct

## Implementation Decisions

attempting to make the code more modular and easier to test in pieces. if we like this pattern (and we don't have to), i'll continue into the rest of the components.

### CODE REVIEW

**Review by mmiermans — APPROVED** [review]
Love extracting code into methods and removing duplicate code, which makes the constructor so much more readable! My comments are optional.

File: src/base/ApplicationAutoscaling.ts [hunk 1]
```diff
@@ -166,27 +173,64 @@ export class ApplicationAutoscaling extends Resource {
...
-        ServiceName: config.ecsServiceName,
-      },
-      alarmActions: [applicationScaleIn.arn],
-      tags: config.tags,
-    });
+  constructor(
+    scope: Construct,
+    name: string,
+    config: ApplicationAutoscalingProps
+  ) {
```
[the remark this answers is not in the document]
Comment (jpetto) [reply src/base/ApplicationAutoscaling.ts:185]: ah, yeah, i'm on board with this. will update.

File: src/base/ApplicationAutoscaling.ts [hunk 2]
```diff
@@ -81,82 +57,113 @@ export class ApplicationAutoscaling extends Resource {
...
-            adjustmentType: 'ChangeInCapacity',
-            cooldown: 60,
-            metricAggregationType: 'Average',
+  static generateAutoSclaingPolicy(
+    construct: Resource,
+    config: ApplicationAutoscalingProps,
+    target: AppautoscalingTarget,
+    type: 'In' | 'Out'
+  ): AppautoscalingPolicy {
+    let stepAdjustment;
```
[the remark this answers is not in the document]
Comment (jpetto) [reply src/base/ApplicationAutoscaling.ts:110]: hm, actually this isn't as straight forward as it seems. the `AppautoscalingPolicyStepScalingPolicyConfigurationStepAdjustment` type doesn't match the code we have. digging into the type file, i see this function immediately after the type definition: `appautoscalingPolicyStepScalingPolicyConfigurationStepAdjustmentToTerraform`. so...i'm going to skip this for now. i don't want to put this level of code change in this PR.

**Comment by jpetto** [comment]
@bassrock @mmiermans could i get another review here after my last commit? i think we're good to roll here.

File: src/base/ApplicationAutoscaling.ts [hunk 3]
```diff
@@ -35,10 +35,69 @@ export class ApplicationAutoscaling extends Resource {
   ) {
     super(scope, name);
 
-    const autoscalingIamRole = new IamRole(this, `autoscaling_role`, {
+    // set up IAM role & policy
+    const iamRole = ApplicationAutoscaling.generateIamRole(this, config);
+    ApplicationAutoscaling.generateIamRolePolicy(this, config, iamRole);
+
+    // set up autoscaling target & in/out policies
+    const autoScalingTarget = ApplicationAutoscaling.generateAutoScalingTarget(
```
Comment (bassrock) [inline src/base/ApplicationAutoscaling.ts:43]: this is nice.

**Review by bassrock — APPROVED** [review]
👍 this looks good to me!
